Dosdebug 22

andrewbird opened the pull request:
Tidy up for dosdebug's value processing mainly including fix for #616

stsp commented on src/tools/debugger/mhpdbgc.c:
```diff
@@ -1511,54 +1588,59 @@ void mhp_cmd(const char * cmd)
+
+    if (decode_symreg(argv[1], &rnum, NULL)) {
+      if (!IN_DPMI) {
+        mhp_printf("register values only valid in DPMI\n");
```
This is quite bad, because unless you
are "stopped", IN_DPMI will change many
times a second. This just is not going to work.
I think we need to remove the reg option.

andrewbird replied:
Several lines above we return if not 'stopped'. does that make it alright, or shall I just rip it out?

stsp replied:
Ah, so now you do entire printing under stop...
Well I don't think this is a good idea because
for example you can dump mem while running -
its all fine unless you start fiddling with regs,
in which case you need this `IN_DPMI` thingy.
So I really think this makes the whole thing
inconsistent and should be removed. We then
don't need "stopped", which is a bit of a relief.

stsp commented on src/tools/debugger/mhpdbgc.c:
```diff
@@ -1511,54 +1588,59 @@ void mhp_cmd(const char * cmd)
+  if (!check_for_stopped())
+    return;
+
+  if (argc > 2 && !getval_ui(argv[2], 0, &lines)) {
```
But IMHO it makes absolutely no sense
to specify the amount of lines together with
the descriptor value to print, as you will
obviously print only one in this case.
I just think the changes around here are an
unnecessary complications w/o real use.

andrewbird replied:
I would agree if I knew for sure how many ldt entries it was possible to print. I looked at [Wikipedia GDT page](https://en.wikipedia.org/wiki/Global_Descriptor_Table) and it suggests 8192 entries. Since we are unlikely to print all in one go we might need some way of starting at an offset. Of course I'm happy to remove the lines arg as I am very unlikely to use it myself, but just conscious of not wrecking it for others.

stsp replied:
8192 is a maximum value.
DPMI progs rarely use more than a few
dozens (well, windows can use much more).
While printing ranges may be useful, printing
a single entry you are interested in, is much,
much more useful when you debug something.
And these are the two entirely different operations
(print range and  print particular entry).
So how about the following:
1. Leave the current logic for the first invocation
2. If user pressed Enter again (repeat last cmd),
then you print next 16 entries no matter how much
you printed on step 1.

?

stsp commented on src/tools/debugger/mhpdbgc.c:
```diff
@@ -114,8 +114,7 @@ char loopbuf[4] = "";
    {"r" ,            mhp_regs},
-   {"e",             mhp_enter},
-   {"ed",            mhp_enter},
+   {"m",             mhp_modmem},
```
Could  you please explain a bit more about
the reason to rename? With what exactly command
was the clash?
Also maybe "poke"? :) Doesn't matter much,
"modmem" can do too.

stsp commented:
Other than a few comments above, looks cool. :)

andrewbird commented:
1/ mhp_print_ldt()
    a) now prints 16 entry page or 1 entry if selector used.
    b) register selector source now removed
    c) check for stopped now removed
2/ mhp_modmem() renamed to mhp_memset()

Hope that's okay now, if not just say?

stsp commented:
But now I simply don't see the patch that
changed print_ldt(). Only some clean-ups
and small tweaks are here. Where is it?
Have you considered "repeat last cmd"
thing, or was it already there?

andrewbird commented:
Okay I'll work on separating it out from the general cleanups.

andrewbird commented:
I've split out the tidyup of the print_ldt function and tweaked it slightly for readability. I only have two ldt entries here so I temporarily changed the page size to 1 and repeated `ldt` commands cycled through fine.

stsp commented:
Thanks!
Will test now. :)

stsp commented:
It actually doesn't print 16 subsequent
lines on Enter repeat. It always prints the
same single line.

andrewbird commented:
After `ldt` or `ldt xx` ? The former should, the second shouldn't.

stsp commented:
After "ldt xx". I was expecting the
subsequent Enter to list subsequent
entries, not the same ones, as that
makes sense.

andrewbird commented:
Mmm I had it working okay before like
`ldt xx`
`ldt`
`<enter>`

but
`ldt xx`
`<enter>`

is much harder as the `<enter>` is doing `ldt xx` not `ldt`. So I changed it to make the two commands independent.

stsp commented:
Maybe at one point we should extend
the API and pass the `repeat` flag that
will indicate whether it is a virgin cmd or
a repeat... Not a big deal of course.
